---
title: "ChIP-nexus Processing Pipeline"
author: "Melanie Weilert"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  rmarkdown::github_document:
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{'Stowers Institute of Medical Research'}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

#Introduction

The purpose of this script is to provide a readable workflow for processing ChIP-nexus data (http://www.nature.com/nbt/journal/v33/n4/full/nbt.3121.html). ChIP‐nexus (Chromatin‐Immunoprecipitation with nucleotide resolution using exonuclease digestion, unique barcode and single ligation) is a ChIP‐exo protocol that makes use of a unique barcode to identify duplicate reads and a novel library preparation strategy that is adopted from iCLIP.

Other alignment tools can be used at the discretion of the investigator, but the preprocessing_fastq.R and the process_bam.R are meant to act as helpers for easy downstream analysis of ChIP-nexus data.

Detailed information on ChIP-nexus experimental protocols can be found at: http://research.stowers.org/zeitlingerlab/protocols.html

#Computational Setup

The following Unix and R tools need to be set up as follows:

## UNIX environment

Before running any of the pipelines below, you will first need to ensure that your UNIX environment is properly configured. The pipelines currently require the following:

 - R 3.2.5
 - Bioconductor 3.2
 - bowtie 1.1.2
 - gzcat
 - cutadapt 1.8.1
 - samtools 1.3.1
 
### bowtie

The pipelines look for bowtie to be installed at `~/apps/bowtie/bowtie`:

```{bash, eval=F}
~/apps/bowtie/bowtie --version
```

If the above command does not work, download and uncompress bowtie 1.1.2 into `~/apps/bowtie`.

Modify your `~/.bashrc` to set the `BOWTIE_INDEXES` environment variable and add `~/bin` to your path.

```{bash, eval=F}
export BOWTIE_INDEXES=[path to bowtie indexes]
export PATH=~/bin:$PATH
```

### cutadapt

The ChIP-nexus pipelines use `cutadapt` to perform adapter trimming. Install version 1.8.1 locally via `pip` and ensure it is in your PATH:

```{bash, eval=F}
pip install --user cutadapt==1.8.1 
~/.local/bin/cutadapt --version
mkdir -p ~/bin
ln -s ~/.local/bin/cutadapt ~/bin
```

### gzcat

The UNIX command `zcat` should be symlinked to `gzcat`. The pipelines use `gzcat` instead of `zcat` for compatability with Mac OS X. 

```{bash, eval=F}
ln -s `which zcat` ~/bin/gzcat
```

### samtools

Verify that `samtools` version 1.2 or higher is available:

```{bash, eval=F}
samtools --version
```

## R and Bioconductor environment

Verify that you have the proper version of R and Bioconductor installed:

```{bash, eval=F}
Rscript --version
Rscript -e "library(BiocInstaller)"
```

Start an R session and verify you have the necessary packages installed:

```{r r_packages}
library(dplyr)
library(pander)
cran_packages <- c("magrittr", "RCurl", "optparse", "stringr", "ascii", "data.table")
bioc_packages <- c("GenomicAlignments", "Rsamtools", "rtracklayer",
                   "GenomicRanges", "chipseq", "ShortRead")
packages <- data_frame(Package = cran_packages, Source = "CRAN") %>%
            rbind(data_frame(Package = bioc_packages, Source = "BioC")) %>%
            mutate(Installed = Package %in% rownames(installed.packages()))
packages %>% pander(caption="Installation status of required packages")
```

##Step 1: Preprocessing FASTQ reads (optional, but recommended)

Given a sequencing file of ChIP-nexus reads, this step removes reads that don't have the proper fixed barcode, moves the random barcode sequences to the FASTQ read name, and removes both the fixed and random barcodes from the reads. Note that this step is highly recommended, but optional. 

Inputs:
-f, --file: Path of ChIP-nexus FASTQ file to process [required]
-o, --output: Output FASTQ file (gzip compressed) [required]

-t, --trim: Pre-trim all reads to this length before processing [default=0]
-k, --keep: Minimum number of bases required after barcode to keep read [default=18]
-b, --barcode: Barcode sequences (comma-separated) that follow random barcode") [default="CTGA"]
-r, --randombarcode: Number of bases at the start of each read used for random barcode [default=5]
-c, --chunksize: Number of reads to process at once (in thousands) [default=1000]
-p, --processors: Number of simultaneous processing cores to utilize [default=2]

Outputs: gzip-compressed FASTQ file with filtered reads

###Example Use Case:

```{bash, eval=F}
Rscript scripts/preprocess_fastq.r -f chipnexus_sample.fastq.gz \
                           -k 22 -b CTGA -r 5 -p 4 -c 1000 \
                           -o chipnexus_sample_processed.fastq.gz
```

##Step 2: Alignment

This step aligns the processed FASTQ file to the genome using bowtie and cutadapt with the appropriate indexing. Manual command line entry or use of a premade script (scripts/align_chipnexus.sh) can be used.

###Example Use Case:

```{bash, eval=F}
#Manual command line
bowtie -S -p 4 --chunkmbs 512 -k 1 -m 1 -v 2 --best --strata \
        bowtie_indexes/dm6 <(cutadapt -m 22 -O 4 -e 0.2 --quiet \
                       -a AGATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC \
                       -a GATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC \
                       -a ATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC \
                       -a TCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC \
                       -a CGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC \
                       chipnexus_sample_processed.fastq.gz) | samtools view -F 4 -Sbo chipnexus_sample.bam -
#Script
scripts/align_chipnexus.sh chipnexus_sample_processed.fastq.gz bowtie_indexes/dm6
```

###Optional: Sort your BAM file for reduced storage and quicker processing downstream.

```{bash, eval=F}
samtools sort chipnexus_sample.bam
```

##Step 3: Process BAM

This step removes reads that align to the same genomic position and have identical random barcodes, resizes the reads to a width of 1 (the first base sequenced), and saves the results as a GRanges object.

Inputs:
-f, --file: Path of BAM file to process [required]
-o, --output: Output FASTQ file (gzip compressed) [required]
-p, --paired: Call option if the alignment was conducted in paired-end mode [default=FALSE]

###Example Use Case:

``` {bash, eval=F}
Rscript process_bam.r -f chipnexus_sample.bam -n chipnexus_sample
```

##Step 4: Generate individual strand BigWigs

This step generates a positive and negative strand BigWig file from the supplied GRanges object.

Inputs:
-f, --file: Path of GRanges file to process [required]
-n, --name: Output prefix for BW file [required]
-m, --normalization: Call option if the output BW file should be normalized to RPM [default=FALSE]

###Example Use Case

``` {bash, eval=F}
Rscript split_granges_by_strand.r -r chipnexus_sample.granges.rds
```

These bigwig files can be used for downstream analysis and loaded into an IGV/UCSC/etc genome browser for survery of results.






